src/http/web_socket.cr

```diff
@@ -37,23 +39,55 @@ class HTTP::WebSocket
     new(Protocol.new(host, path, port, tls))
   end
 
+  def on_ping(&@on_ping : String ->)
+  end
+
+  def on_pong(&@on_pong : String ->)
+  end
+
   def on_message(&@on_message : String ->)
   end
 
   def on_close(&@on_close : String ->)
   end
 
+  protected def check_open
+    raise IO::Error.new "closed socket" if closed?
+  end
+
   def send(message)
+    check_open
     @ws.send(message)
   end
 
+  # It's possible to send a PING frame, which the client must respond to
+  # with a PONG, or the server can send an unsolicited PONG frame
+  # which the client should not respond to.
+  #
+  # See `#pong`.
+  def ping(message = nil)
+    check_open
+    @ws.ping(message)
+  end
+
+  # Server can send an unsolicited PONG frame which the client should not respond to.
+  #
+  # See `#ping`.
+  def pong(message = nil)
+    check_open
+    @ws.pong(message)
+  end
+
   def stream(binary = true, frame_size = 1024)
+    check_open
     @ws.stream(binary: binary, frame_size: frame_size) do |io|
       yield io
     end
   end
 
   def close(message = nil)
+    check_open
```

Comment:
Double closing a websocket shouldn't error. This is how every IO works right now and we prefer this behaviour as no bug can result from closing a closed socket. So this should `return if @closed` as the first thing.